### PR TITLE
Migrate component isolation away from multiprocessing

### PR DESCRIPTION
**Pull request by pipermerriam**
### What was wrong?

Multiprocessing isn't async friendly.

### How was it fixed?

Convert `AsyncioIsolatedComponent` to use `asyncio-run-in-process` for process isolation.

#### Cute Animal Picture

Adorable two-week-old seal pups show playful side under watchful eye of mother 3

### CODE REVIEW

**Comment by pipermerriam** [comment]
Depends on https://github.com/ethereum/asyncio-run-in-process/pull/5 because `uvloop + pytest + async_run_in_process` aren't getting along due to `uvloop` not liking pytest's mocked out stdout/err/in capturing which I wasn't able to get around locally.

**Comment by pipermerriam** [comment]
currently blocked by https://github.com/ethereum/asyncio-run-in-process/pull/6

File: tests/core/components/test_isolated_component.py [hunk 1]
```diff
@@ -99,7 +48,7 @@ def log_listener(trinity_config):
 
 @pytest.mark.asyncio
 async def test_asyncio_isolated_component(boot_info,
-                                          log_listener):
+                                          log_listener, event_loop):
```
Comment (pipermerriam) [inline tests/core/components/test_isolated_component.py]: looks like I can remove the `event_loop`

File: trinity/extensibility/asyncio.py [hunk 2]
```diff
@@ -1,79 +1,67 @@
...
 
+import logging
+logger = logging.getLogger('trinity')
+
 
 class AsyncioIsolatedComponent(BaseIsolatedComponent):
-    async def run(self) -> None:
-        """
-        Call chain is:
+    def get_subprocess_kwargs(self) -> Optional[SubprocessKwargs]:
```
Comment (pipermerriam) [inline trinity/extensibility/asyncio.py]: This API is really only currently used for testing but it's needed to get around the unfortunate behavior of `asyncio.open_subprocess_exc + pytest + uvloop`.  Pytest replaces `sys.stdin/out/err` with versions that don't have a `fileno` attribute which `uvloop` gets very unhappy about when you try to open a subprocess.

File: trinity/extensibility/asyncio.py [hunk 3]
```diff
@@ -1,79 +1,67 @@
...
-        loop.add_signal_handler(signal.SIGINT, task.cancel)
-        loop.add_signal_handler(signal.SIGTERM, task.cancel)
-        try:
-            loop.run_until_complete(task)
-        except asyncio.CancelledError:
-            pass
-        finally:
-            loop.close()
+                await proc.wait()
+            except asyncio.CancelledError as err:
```
Comment (pipermerriam) [inline trinity/extensibility/asyncio.py]: realizing this whole `try/except` can go away as this behavior is now actually native to `asyncio-run-in-process`

File: trinity/extensibility/component_manager.py [hunk 4]
```diff
@@ -52,61 +54,62 @@ def __init__(self,
...
+                        '/'.join(component.name for component in enabled_components),
+                    )
+                    # Concurrently start the components.
+                    await asyncio.gather(*(
+                        stack.enter_async_context(run_component(component))
+                        for component in enabled_components
+                    ))
+                    self.logger.info("Components started")
+                    try:
+                        await self._trigger_component_exit.wait()
```
Comment (pipermerriam) [inline trinity/extensibility/component_manager.py]: This mechanism is still not ideal or bulletproof but its working for now.  I intend to continue iterating on this.

File: trinity/sync/beam/chain.py [hunk 5]
```diff
@@ -191,7 +191,7 @@ def __init__(
                 "Timed out while trying to fulfill prerequisites of "
                 f"sync launch strategy: {exc} from {self._launch_strategy}"
             )
-            await self.cancel()
+            self.cancel_nowait()
```
Comment (pipermerriam) [inline trinity/sync/beam/chain.py]: I'm not sure how much this matters but the way that `p2p.service.BaseService` can end up *waiting* for things to shutdown seemed to be causing problems.

File: trinity/extensibility/asyncio.py [hunk 6]
```diff
@@ -1,79 +1,68 @@
...
+                proc.send_signal(signal.SIGINT)
+                try:
+                    await asyncio.wait_for(proc.wait(), timeout=2)
+                except asyncio.TimeoutError:
+                    pass
```
Comment (carver) [inline trinity/extensibility/asyncio.py]: Can we log that the process didn't shut down cleanly/quickly here? Or is it too tricky b/c the logger process might be dead.

Comment (pipermerriam) [reply trinity/extensibility/asyncio.py]: No, we should be able to log here, I'll add it.

File: trinity/extensibility/asyncio.py [hunk 7]
```diff
@@ -1,79 +1,68 @@
...
+logger = logging.getLogger('trinity')
+
 
 class AsyncioIsolatedComponent(BaseIsolatedComponent):
-    async def run(self) -> None:
-        """
-        Call chain is:
+    def get_subprocess_kwargs(self) -> Optional[SubprocessKwargs]:
+        # Note that this method currently only exist to facilitate testing.
+        return None
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply trinity/extensibility/asyncio.py]: No, the unset case isn't problematic.  `None` is the default for this argument so it's fine to have the `Optional[..]` typing.

**Pull request merged**